src/components/application_manager/src/resumption/resumption_data_processor.cc

```diff
@@ -0,0 +1,1125 @@
+ list of conditions and the following disclaimer.
+ Redistributions in binary form must reproduce the above copyright notice,
+ this list of conditions and the following
+ disclaimer in the documentation and/or other materials provided with the
+ distribution.
+ Neither the name of the Ford Motor Company nor the names of its contributors
+ may be used to endorse or promote products derived from this software
+ without specific prior written permission.
+ THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <algorithm>
+
+#include "application_manager/application_manager.h"
+#include "application_manager/commands/command_impl.h"
+#include "application_manager/display_capabilities_builder.h"
+#include "application_manager/event_engine/event_observer.h"
+#include "application_manager/message_helper.h"
+#include "application_manager/resumption/resumption_data_processor.h"
+#include "application_manager/smart_object_keys.h"
+
+namespace resumption {
+
+using app_mngr::AppFile;
+using app_mngr::ApplicationSharedPtr;
+using app_mngr::ButtonSubscriptions;
+using app_mngr::ChoiceSetMap;
+using app_mngr::MessageHelper;
+namespace strings = app_mngr::strings;
+namespace event_engine = app_mngr::event_engine;
+namespace commands = app_mngr::commands;
+
+CREATE_LOGGERPTR_GLOBAL(logger_, "Resumption")
+
+bool ResumptionRequestID::operator<(const ResumptionRequestID& other) const {
+  return correlation_id < other.correlation_id ||
+         function_id < other.function_id;
+}
+
+ResumptionDataProcessor::ResumptionDataProcessor(
+    app_mngr::ApplicationManager& application_manager)
+    : event_engine::EventObserver(application_manager.event_dispatcher())
+    , application_manager_(application_manager)
+    , resumption_status_lock_()
+    , register_callbacks_lock_()
+    , request_app_ids_lock_() {}
+
+ResumptionDataProcessor::~ResumptionDataProcessor() {}
+
+void ResumptionDataProcessor::Restore(ApplicationSharedPtr application,
+                                      smart_objects::SmartObject& saved_app,
+                                      ResumeCtrl::ResumptionCallBack callback) {
+  LOG4CXX_AUTO_TRACE(logger_);
+
+  if (!HasDataToRestore(saved_app) &&
```

Comment:
Would it make sense to check `saved_app[strings::application_files]` here as well or will that not exist without another element?

Reply:
@iCollin yes, it make sense. Even though `AddFiles` function does not trigger any requests to be sent to HMI, this function should be called just to restore internal app information about persisted files and their location. Check has been added in https://github.com/smartdevicelink/sdl_core/pull/3475/commits/8fa02f00fc451a61acdc1f76ea223094333fbfa5